File: agentos_examples/cplusplus/skill_contrl.cpp
#include <ros/ros.h>
#include <actionlib/client/simple_action_client.h>
#include <agent_msgs/ControlAction.h>

// Feedback callback function
void feedbackCallback(const agent_msgs::ControlFeedbackConstPtr &feedback)
{
  ROS_INFO("Feedback progress: %f", feedback->progress);
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "skill_control_test");
  ros::NodeHandle nh;

  std::string arg_skill_name = "voice_interaction";
  int arg_command = 1;

  // Create the action client
  actionlib::SimpleActionClient<agent_msgs::ControlAction> ac("/agent_skill/" + arg_skill_name + "/control", true);

  ROS_INFO("Waiting for action server to start...");
  if (!ac.waitForServer(ros::Duration(3.0)))
  {
    ROS_ERROR("Action server not available after waiting.");
    return 1;
  }

  // Creating a goal
  agent_msgs::ControlGoal goal;
  goal.command = arg_command;

  // Send goal to the server with feedback callback
  try
  {
    ac.sendGoal(goal, actionlib::SimpleActionClient<agent_msgs::ControlAction>::SimpleDoneCallback(),
                actionlib::SimpleActionClient<agent_msgs::ControlAction>::SimpleActiveCallback(),
                &feedbackCallback);
  }
  catch (const ros::Exception &e)
  {
    ROS_ERROR("Failed to send goal: %s", e.what());
    return 1;
  }

  ROS_INFO("Waiting for result...");
  if (!ac.waitForResult(ros::Duration(3.0)))
  {
    ROS_ERROR("Action did not finish before the timeout.");
    return 1;
  }

  // Get and print the result
  agent_msgs::ControlResultConstPtr result = ac.getResult();
  ROS_INFO("Result: %d", result->result);

  return 0;
}

File: agentos_examples/python/skill_contrl.py
#! /usr/bin/env python3

import rospy
import actionlib

from agent_msgs.msg import ControlAction, ControlGoal


def feedback_callback(feedback):
  rospy.loginfo(f"Feedback progress: {feedback.progress}")


if __name__ == '__main__':
  rospy.init_node('skill_control_test')

  arg_skill_name = 'voice_interaction'
  arg_command = 1

  cli = actionlib.SimpleActionClient(f"/agent_skill/{arg_skill_name}/control", ControlAction)

  rospy.loginfo("Waiting for action server to start...")
  if not cli.wait_for_server(timeout=rospy.Duration(3.0)):
    rospy.logerr("Action server not available after waiting.")
    exit(1)

  goal = ControlGoal(command=arg_command)

  try:
    cli.send_goal(goal, feedback_cb=feedback_callback)
  except rospy.ROSException as e:
    rospy.logerr(f"Failed to send goal: {e}")
    exit(1)


  rospy.loginfo("Waiting for result...")
  if not cli.wait_for_result(timeout=rospy.Duration(3.0)):
    rospy.logerr("Action did not finish before the timeout.")
    exit(1)

  result = cli.get_result()
  rospy.loginfo(f"Result result: {result.result}")

